src/main/java/com/artozersky/HackerNewsAPI/service/impl/NewsPostServiceImpl.java

```diff
@@ -0,0 +1,262 @@
+
+    System.out.println("fetched from db: " + dbPosts.size() );
+
+    List<NewsPostModelImpl> combinedPosts = new ArrayList<>(cachedPosts);
+    combinedPosts.addAll(dbPosts);
+
+    return combinedPosts.stream()
+        .map(this::convertToDTO)
+        .collect(Collectors.toList());
+}
+
+    
+    @Override
+    public NewsPostsResponseDTOImpl savePost(@Valid NewsPostsCreateDTOImpl postCreateDTO) {
+        
+        NewsPostModelImpl post = modelMapper.map(postCreateDTO, NewsPostModelImpl.class);
+        
+        post.initialize();
+        
+        NewsPostModelImpl savedPost = postRepository.save(post);
+        
+        NewsPostsResponseDTOImpl responseDTO = modelMapper.map(savedPost, NewsPostsResponseDTOImpl.class);
+        
+        responseDTO.setMessage("Post created successfully");
+
+        updateCacheWithTopPosts();
+        
+        return responseDTO;
+    }
+
+    @Override
+    public NewsPostsResponseDTOImpl updatePost(NewsPostsUpdateDTOImpl postUpdateDTO, Long postId) {
+        
+        NewsPostModelImpl post = postRepository.findById(postId)
+        .orElseThrow(() -> new CustomNotFoundException("Post not found with id: " + postId));
+        
+        boolean isRedundant = true;
+        
+        if (postUpdateDTO.getTitle() != null && !postUpdateDTO.getTitle().trim().isEmpty()) {
+            if (!postUpdateDTO.getTitle().equals(post.getTitle())) {
+                post.setTitle(postUpdateDTO.getTitle());
+                isRedundant = false;
+            }
+        }
+        
+        if (postUpdateDTO.getUrl() != null && !postUpdateDTO.getUrl().trim().isEmpty()) {
+            if (!postUpdateDTO.getUrl().equals(post.getUrl())) {
+                post.setUrl(postUpdateDTO.getUrl());
+                isRedundant = false;
+            }
+        }
+        
+        if(isRedundant) {
+            NewsPostsResponseDTOImpl responseDTO = modelMapper.map(post, NewsPostsResponseDTOImpl.class);
+            responseDTO.setMessage("No changes detected. Update skipped.");
+            return responseDTO;
+        }
+
+        post.onPostUpdate();
+        
+        NewsPostModelImpl updatedPost = postRepository.save(post);
+        
+        updateCacheWithTopPosts();
+
+        NewsPostsResponseDTOImpl responseDTO = modelMapper.map(updatedPost, NewsPostsResponseDTOImpl.class);
+        responseDTO.setMessage("Post updated successfully");
+
+        
+        return responseDTO;
+    }
+    
+    @Override
+    public NewsPostsResponseDTOImpl upVote(Long id) {
+        NewsPostModelImpl post = postRepository.findById(id)
+        .orElseThrow(() -> new CustomNotFoundException("Post not found with id: " + id));
+        post.upVote();        
+        NewsPostModelImpl updatedPost = postRepository.save(post);
+
+        //if in the cache update the cache
+        if (cacheService.getPostFromCacheById(id) != null) {
+            cacheService.putPostInCache(updatedPost);
```

Comment:
is this put post will put the post in the right order and in the right place in the cache according to its updated rank? (upvotes/downvotes effect rank so they effect the top posts ordering).

Reply:
now it will